// renovate/dist/modules/manager/terragrunt/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.categories = exports.defaultConfig = exports.supportsLockFileMaintenance = exports.supportedDatasources = exports.extractPackageFile = exports.updateArtifacts = void 0;
const git_tags_1 = require("../../datasource/git-tags");
const github_tags_1 = require("../../datasource/github-tags");
const terraform_module_1 = require("../../datasource/terraform-module");
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
exports.supportedDatasources = [
    git_tags_1.GitTagsDatasource.id,
    github_tags_1.GithubTagsDatasource.id,
    terraform_module_1.TerraformModuleDatasource.id,
];
exports.supportsLockFileMaintenance = true;
exports.defaultConfig = {
    commitMessageTopic: 'Terragrunt dependency {{depName}}',
    fileMatch: ['(^|/)terragrunt\\.hcl$'],
};
exports.categories = ['iac', 'terraform'];
