// @aws-sdk/client-codecommit/dist-cjs/CodeCommit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeCommit = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const CodeCommitClient_1 = require("./CodeCommitClient");
const AssociateApprovalRuleTemplateWithRepositoryCommand_1 = require("./commands/AssociateApprovalRuleTemplateWithRepositoryCommand");
const BatchAssociateApprovalRuleTemplateWithRepositoriesCommand_1 = require("./commands/BatchAssociateApprovalRuleTemplateWithRepositoriesCommand");
const BatchDescribeMergeConflictsCommand_1 = require("./commands/BatchDescribeMergeConflictsCommand");
const BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand_1 = require("./commands/BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand");
const BatchGetCommitsCommand_1 = require("./commands/BatchGetCommitsCommand");
const BatchGetRepositoriesCommand_1 = require("./commands/BatchGetRepositoriesCommand");
const CreateApprovalRuleTemplateCommand_1 = require("./commands/CreateApprovalRuleTemplateCommand");
const CreateBranchCommand_1 = require("./commands/CreateBranchCommand");
const CreateCommitCommand_1 = require("./commands/CreateCommitCommand");
const CreatePullRequestApprovalRuleCommand_1 = require("./commands/CreatePullRequestApprovalRuleCommand");
const CreatePullRequestCommand_1 = require("./commands/CreatePullRequestCommand");
const CreateRepositoryCommand_1 = require("./commands/CreateRepositoryCommand");
const CreateUnreferencedMergeCommitCommand_1 = require("./commands/CreateUnreferencedMergeCommitCommand");
const DeleteApprovalRuleTemplateCommand_1 = require("./commands/DeleteApprovalRuleTemplateCommand");
const DeleteBranchCommand_1 = require("./commands/DeleteBranchCommand");
const DeleteCommentContentCommand_1 = require("./commands/DeleteCommentContentCommand");
const DeleteFileCommand_1 = require("./commands/DeleteFileCommand");
const DeletePullRequestApprovalRuleCommand_1 = require("./commands/DeletePullRequestApprovalRuleCommand");
const DeleteRepositoryCommand_1 = require("./commands/DeleteRepositoryCommand");
const DescribeMergeConflictsCommand_1 = require("./commands/DescribeMergeConflictsCommand");
const DescribePullRequestEventsCommand_1 = require("./commands/DescribePullRequestEventsCommand");
const DisassociateApprovalRuleTemplateFromRepositoryCommand_1 = require("./commands/DisassociateApprovalRuleTemplateFromRepositoryCommand");
const EvaluatePullRequestApprovalRulesCommand_1 = require("./commands/EvaluatePullRequestApprovalRulesCommand");
const GetApprovalRuleTemplateCommand_1 = require("./commands/GetApprovalRuleTemplateCommand");
const GetBlobCommand_1 = require("./commands/GetBlobCommand");
const GetBranchCommand_1 = require("./commands/GetBranchCommand");
const GetCommentCommand_1 = require("./commands/GetCommentCommand");
const GetCommentReactionsCommand_1 = require("./commands/GetCommentReactionsCommand");
const GetCommentsForComparedCommitCommand_1 = require("./commands/GetCommentsForComparedCommitCommand");
const GetCommentsForPullRequestCommand_1 = require("./commands/GetCommentsForPullRequestCommand");
const GetCommitCommand_1 = require("./commands/GetCommitCommand");
const GetDifferencesCommand_1 = require("./commands/GetDifferencesCommand");
const GetFileCommand_1 = require("./commands/GetFileCommand");
const GetFolderCommand_1 = require("./commands/GetFolderCommand");
const GetMergeCommitCommand_1 = require("./commands/GetMergeCommitCommand");
const GetMergeConflictsCommand_1 = require("./commands/GetMergeConflictsCommand");
const GetMergeOptionsCommand_1 = require("./commands/GetMergeOptionsCommand");
const GetPullRequestApprovalStatesCommand_1 = require("./commands/GetPullRequestApprovalStatesCommand");
const GetPullRequestCommand_1 = require("./commands/GetPullRequestCommand");
const GetPullRequestOverrideStateCommand_1 = require("./commands/GetPullRequestOverrideStateCommand");
const GetRepositoryCommand_1 = require("./commands/GetRepositoryCommand");
const GetRepositoryTriggersCommand_1 = require("./commands/GetRepositoryTriggersCommand");
const ListApprovalRuleTemplatesCommand_1 = require("./commands/ListApprovalRuleTemplatesCommand");
const ListAssociatedApprovalRuleTemplatesForRepositoryCommand_1 = require("./commands/ListAssociatedApprovalRuleTemplatesForRepositoryCommand");
const ListBranchesCommand_1 = require("./commands/ListBranchesCommand");
const ListPullRequestsCommand_1 = require("./commands/ListPullRequestsCommand");
const ListRepositoriesCommand_1 = require("./commands/ListRepositoriesCommand");
const ListRepositoriesForApprovalRuleTemplateCommand_1 = require("./commands/ListRepositoriesForApprovalRuleTemplateCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const MergeBranchesByFastForwardCommand_1 = require("./commands/MergeBranchesByFastForwardCommand");
const MergeBranchesBySquashCommand_1 = require("./commands/MergeBranchesBySquashCommand");
const MergeBranchesByThreeWayCommand_1 = require("./commands/MergeBranchesByThreeWayCommand");
const MergePullRequestByFastForwardCommand_1 = require("./commands/MergePullRequestByFastForwardCommand");
const MergePullRequestBySquashCommand_1 = require("./commands/MergePullRequestBySquashCommand");
const MergePullRequestByThreeWayCommand_1 = require("./commands/MergePullRequestByThreeWayCommand");
const OverridePullRequestApprovalRulesCommand_1 = require("./commands/OverridePullRequestApprovalRulesCommand");
const PostCommentForComparedCommitCommand_1 = require("./commands/PostCommentForComparedCommitCommand");
const PostCommentForPullRequestCommand_1 = require("./commands/PostCommentForPullRequestCommand");
const PostCommentReplyCommand_1 = require("./commands/PostCommentReplyCommand");
const PutCommentReactionCommand_1 = require("./commands/PutCommentReactionCommand");
const PutFileCommand_1 = require("./commands/PutFileCommand");
const PutRepositoryTriggersCommand_1 = require("./commands/PutRepositoryTriggersCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const TestRepositoryTriggersCommand_1 = require("./commands/TestRepositoryTriggersCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateApprovalRuleTemplateContentCommand_1 = require("./commands/UpdateApprovalRuleTemplateContentCommand");
const UpdateApprovalRuleTemplateDescriptionCommand_1 = require("./commands/UpdateApprovalRuleTemplateDescriptionCommand");
const UpdateApprovalRuleTemplateNameCommand_1 = require("./commands/UpdateApprovalRuleTemplateNameCommand");
const UpdateCommentCommand_1 = require("./commands/UpdateCommentCommand");
const UpdateDefaultBranchCommand_1 = require("./commands/UpdateDefaultBranchCommand");
const UpdatePullRequestApprovalRuleContentCommand_1 = require("./commands/UpdatePullRequestApprovalRuleContentCommand");
const UpdatePullRequestApprovalStateCommand_1 = require("./commands/UpdatePullRequestApprovalStateCommand");
const UpdatePullRequestDescriptionCommand_1 = require("./commands/UpdatePullRequestDescriptionCommand");
const UpdatePullRequestStatusCommand_1 = require("./commands/UpdatePullRequestStatusCommand");
const UpdatePullRequestTitleCommand_1 = require("./commands/UpdatePullRequestTitleCommand");
const UpdateRepositoryDescriptionCommand_1 = require("./commands/UpdateRepositoryDescriptionCommand");
const UpdateRepositoryNameCommand_1 = require("./commands/UpdateRepositoryNameCommand");
const commands = {
    AssociateApprovalRuleTemplateWithRepositoryCommand: AssociateApprovalRuleTemplateWithRepositoryCommand_1.AssociateApprovalRuleTemplateWithRepositoryCommand,
    BatchAssociateApprovalRuleTemplateWithRepositoriesCommand: BatchAssociateApprovalRuleTemplateWithRepositoriesCommand_1.BatchAssociateApprovalRuleTemplateWithRepositoriesCommand,
    BatchDescribeMergeConflictsCommand: BatchDescribeMergeConflictsCommand_1.BatchDescribeMergeConflictsCommand,
    BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand: BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand_1.BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand,
    BatchGetCommitsCommand: BatchGetCommitsCommand_1.BatchGetCommitsCommand,
    BatchGetRepositoriesCommand: BatchGetRepositoriesCommand_1.BatchGetRepositoriesCommand,
    CreateApprovalRuleTemplateCommand: CreateApprovalRuleTemplateCommand_1.CreateApprovalRuleTemplateCommand,
    CreateBranchCommand: CreateBranchCommand_1.CreateBranchCommand,
    CreateCommitCommand: CreateCommitCommand_1.CreateCommitCommand,
    CreatePullRequestCommand: CreatePullRequestCommand_1.CreatePullRequestCommand,
    CreatePullRequestApprovalRuleCommand: CreatePullRequestApprovalRuleCommand_1.CreatePullRequestApprovalRuleCommand,
    CreateRepositoryCommand: CreateRepositoryCommand_1.CreateRepositoryCommand,
    CreateUnreferencedMergeCommitCommand: CreateUnreferencedMergeCommitCommand_1.CreateUnreferencedMergeCommitCommand,
    DeleteApprovalRuleTemplateCommand: DeleteApprovalRuleTemplateCommand_1.DeleteApprovalRuleTemplateCommand,
    DeleteBranchCommand: DeleteBranchCommand_1.DeleteBranchCommand,
    DeleteCommentContentCommand: DeleteCommentContentCommand_1.DeleteCommentContentCommand,
    DeleteFileCommand: DeleteFileCommand_1.DeleteFileCommand,
    DeletePullRequestApprovalRuleCommand: DeletePullRequestApprovalRuleCommand_1.DeletePullRequestApprovalRuleCommand,
    DeleteRepositoryCommand: DeleteRepositoryCommand_1.DeleteRepositoryCommand,
    DescribeMergeConflictsCommand: DescribeMergeConflictsCommand_1.DescribeMergeConflictsCommand,
    DescribePullRequestEventsCommand: DescribePullRequestEventsCommand_1.DescribePullRequestEventsCommand,
    DisassociateApprovalRuleTemplateFromRepositoryCommand: DisassociateApprovalRuleTemplateFromRepositoryCommand_1.DisassociateApprovalRuleTemplateFromRepositoryCommand,
    EvaluatePullRequestApprovalRulesCommand: EvaluatePullRequestApprovalRulesCommand_1.EvaluatePullRequestApprovalRulesCommand,
    GetApprovalRuleTemplateCommand: GetApprovalRuleTemplateCommand_1.GetApprovalRuleTemplateCommand,
    GetBlobCommand: GetBlobCommand_1.GetBlobCommand,
    GetBranchCommand: GetBranchCommand_1.GetBranchCommand,
    GetCommentCommand: GetCommentCommand_1.GetCommentCommand,
    GetCommentReactionsCommand: GetCommentReactionsCommand_1.GetCommentReactionsCommand,
    GetCommentsForComparedCommitCommand: GetCommentsForComparedCommitCommand_1.GetCommentsForComparedCommitCommand,
    GetCommentsForPullRequestCommand: GetCommentsForPullRequestCommand_1.GetCommentsForPullRequestCommand,
    GetCommitCommand: GetCommitCommand_1.GetCommitCommand,
    GetDifferencesCommand: GetDifferencesCommand_1.GetDifferencesCommand,
    GetFileCommand: GetFileCommand_1.GetFileCommand,
    GetFolderCommand: GetFolderCommand_1.GetFolderCommand,
    GetMergeCommitCommand: GetMergeCommitCommand_1.GetMergeCommitCommand,
    GetMergeConflictsCommand: GetMergeConflictsCommand_1.GetMergeConflictsCommand,
    GetMergeOptionsCommand: GetMergeOptionsCommand_1.GetMergeOptionsCommand,
    GetPullRequestCommand: GetPullRequestCommand_1.GetPullRequestCommand,
    GetPullRequestApprovalStatesCommand: GetPullRequestApprovalStatesCommand_1.GetPullRequestApprovalStatesCommand,
    GetPullRequestOverrideStateCommand: GetPullRequestOverrideStateCommand_1.GetPullRequestOverrideStateCommand,
    GetRepositoryCommand: GetRepositoryCommand_1.GetRepositoryCommand,
    GetRepositoryTriggersCommand: GetRepositoryTriggersCommand_1.GetRepositoryTriggersCommand,
    ListApprovalRuleTemplatesCommand: ListApprovalRuleTemplatesCommand_1.ListApprovalRuleTemplatesCommand,
    ListAssociatedApprovalRuleTemplatesForRepositoryCommand: ListAssociatedApprovalRuleTemplatesForRepositoryCommand_1.ListAssociatedApprovalRuleTemplatesForRepositoryCommand,
    ListBranchesCommand: ListBranchesCommand_1.ListBranchesCommand,
    ListPullRequestsCommand: ListPullRequestsCommand_1.ListPullRequestsCommand,
    ListRepositoriesCommand: ListRepositoriesCommand_1.ListRepositoriesCommand,
    ListRepositoriesForApprovalRuleTemplateCommand: ListRepositoriesForApprovalRuleTemplateCommand_1.ListRepositoriesForApprovalRuleTemplateCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    MergeBranchesByFastForwardCommand: MergeBranchesByFastForwardCommand_1.MergeBranchesByFastForwardCommand,
    MergeBranchesBySquashCommand: MergeBranchesBySquashCommand_1.MergeBranchesBySquashCommand,
    MergeBranchesByThreeWayCommand: MergeBranchesByThreeWayCommand_1.MergeBranchesByThreeWayCommand,
    MergePullRequestByFastForwardCommand: MergePullRequestByFastForwardCommand_1.MergePullRequestByFastForwardCommand,
    MergePullRequestBySquashCommand: MergePullRequestBySquashCommand_1.MergePullRequestBySquashCommand,
    MergePullRequestByThreeWayCommand: MergePullRequestByThreeWayCommand_1.MergePullRequestByThreeWayCommand,
    OverridePullRequestApprovalRulesCommand: OverridePullRequestApprovalRulesCommand_1.OverridePullRequestApprovalRulesCommand,
    PostCommentForComparedCommitCommand: PostCommentForComparedCommitCommand_1.PostCommentForComparedCommitCommand,
    PostCommentForPullRequestCommand: PostCommentForPullRequestCommand_1.PostCommentForPullRequestCommand,
    PostCommentReplyCommand: PostCommentReplyCommand_1.PostCommentReplyCommand,
    PutCommentReactionCommand: PutCommentReactionCommand_1.PutCommentReactionCommand,
    PutFileCommand: PutFileCommand_1.PutFileCommand,
    PutRepositoryTriggersCommand: PutRepositoryTriggersCommand_1.PutRepositoryTriggersCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    TestRepositoryTriggersCommand: TestRepositoryTriggersCommand_1.TestRepositoryTriggersCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateApprovalRuleTemplateContentCommand: UpdateApprovalRuleTemplateContentCommand_1.UpdateApprovalRuleTemplateContentCommand,
    UpdateApprovalRuleTemplateDescriptionCommand: UpdateApprovalRuleTemplateDescriptionCommand_1.UpdateApprovalRuleTemplateDescriptionCommand,
    UpdateApprovalRuleTemplateNameCommand: UpdateApprovalRuleTemplateNameCommand_1.UpdateApprovalRuleTemplateNameCommand,
    UpdateCommentCommand: UpdateCommentCommand_1.UpdateCommentCommand,
    UpdateDefaultBranchCommand: UpdateDefaultBranchCommand_1.UpdateDefaultBranchCommand,
    UpdatePullRequestApprovalRuleContentCommand: UpdatePullRequestApprovalRuleContentCommand_1.UpdatePullRequestApprovalRuleContentCommand,
    UpdatePullRequestApprovalStateCommand: UpdatePullRequestApprovalStateCommand_1.UpdatePullRequestApprovalStateCommand,
    UpdatePullRequestDescriptionCommand: UpdatePullRequestDescriptionCommand_1.UpdatePullRequestDescriptionCommand,
    UpdatePullRequestStatusCommand: UpdatePullRequestStatusCommand_1.UpdatePullRequestStatusCommand,
    UpdatePullRequestTitleCommand: UpdatePullRequestTitleCommand_1.UpdatePullRequestTitleCommand,
    UpdateRepositoryDescriptionCommand: UpdateRepositoryDescriptionCommand_1.UpdateRepositoryDescriptionCommand,
    UpdateRepositoryNameCommand: UpdateRepositoryNameCommand_1.UpdateRepositoryNameCommand,
};
class CodeCommit extends CodeCommitClient_1.CodeCommitClient {
}
exports.CodeCommit = CodeCommit;
(0, smithy_client_1.createAggregatedClient)(commands, CodeCommit);
