// renovate/dist/config/presets/internal/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.presets = void 0;
/* eslint sort-keys: ["error", "asc", {caseSensitive: false, natural: true}] */
exports.presets = {
    disableTypesNodeMajor: {
        description: 'Disable `major` updates to `@types/node`.',
        packageRules: [
            {
                enabled: false,
                matchPackageNames: ['@types/node'],
                matchUpdateTypes: ['major'],
            },
        ],
    },
    followTypescriptNext: {
        description: 'Keep `typescript` version in sync with the `next` tag.',
        extends: [':followTag(typescript, next)'],
    },
    followTypescriptRc: {
        description: 'Keep `typescript` version in sync with the `rc` tag.',
        extends: [':followTag(typescript, rc)'],
    },
    pinGitHubActionDigests: {
        description: 'Pin `github-action` digests.',
        packageRules: [
            {
                matchDepTypes: ['action'],
                pinDigests: true,
            },
        ],
    },
};
