// renovate/dist/config/presets/internal/merge-confidence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.presets = void 0;
exports.presets = {
    'all-badges': {
        description: 'Show all Merge Confidence badges for pull requests.',
        packageRules: [
            {
                matchDatasources: ['maven', 'npm', 'pypi'],
                matchUpdateTypes: ['patch', 'minor', 'major'],
                prBodyColumns: [
                    'Package',
                    'Change',
                    'Age',
                    'Adoption',
                    'Passing',
                    'Confidence',
                ],
            },
        ],
    },
    'age-confidence-badges': {
        description: 'Show only the Age and Confidence Merge Confidence badges for pull requests.',
        packageRules: [
            {
                matchDatasources: ['maven', 'npm', 'pypi'],
                matchUpdateTypes: ['patch', 'minor', 'major'],
                prBodyColumns: ['Package', 'Change', 'Age', 'Confidence'],
            },
        ],
    },
};
