// renovate/dist/modules/manager/pep621/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PyProjectSchema = void 0;
const zod_1 = require("zod");
const DependencyListSchema = zod_1.z.array(zod_1.z.string()).optional();
const DependencyRecordSchema = zod_1.z
    .record(zod_1.z.string(), zod_1.z.array(zod_1.z.string()))
    .optional();
exports.PyProjectSchema = zod_1.z.object({
    project: zod_1.z
        .object({
        'requires-python': zod_1.z.string().optional(),
        dependencies: DependencyListSchema,
        'optional-dependencies': DependencyRecordSchema,
    })
        .optional(),
    tool: zod_1.z
        .object({
        pdm: zod_1.z
            .object({
            'dev-dependencies': DependencyRecordSchema,
            source: zod_1.z
                .array(zod_1.z.object({
                url: zod_1.z.string(),
                name: zod_1.z.string(),
                verify_ssl: zod_1.z.boolean().optional(),
            }))
                .optional(),
        })
            .optional(),
    })
        .optional(),
});
