// renovate/dist/modules/manager/cargo/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedDatasources = exports.categories = exports.defaultConfig = exports.updateArtifacts = exports.extractPackageFile = exports.supportsLockFileMaintenance = exports.getRangeStrategy = void 0;
const tslib_1 = require("tslib");
const crate_1 = require("../../datasource/crate");
const cargoVersioning = tslib_1.__importStar(require("../../versioning/cargo"));
const artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
const extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
var range_1 = require("./range");
Object.defineProperty(exports, "getRangeStrategy", { enumerable: true, get: function () { return range_1.getRangeStrategy; } });
exports.supportsLockFileMaintenance = true;
exports.defaultConfig = {
    commitMessageTopic: 'Rust crate {{depName}}',
    fileMatch: ['(^|/)Cargo\\.toml$'],
    versioning: cargoVersioning.id,
};
exports.categories = ['rust'];
exports.supportedDatasources = [crate_1.CrateDatasource.id];
