// renovate/dist/modules/manager/npm/extract/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageLock = exports.PackageLockPreV3Schema = exports.PackageLockV3Schema = void 0;
const zod_1 = require("zod");
const schema_utils_1 = require("../../../../util/schema-utils");
exports.PackageLockV3Schema = zod_1.z.object({
    lockfileVersion: zod_1.z.literal(3),
    packages: (0, schema_utils_1.LooseRecord)(zod_1.z
        .string()
        .transform((x) => x.replace(/^node_modules\//, ''))
        .refine((x) => x.trim() !== ''), zod_1.z.object({ version: zod_1.z.string() })),
});
exports.PackageLockPreV3Schema = zod_1.z
    .object({
    lockfileVersion: zod_1.z.union([zod_1.z.literal(2), zod_1.z.literal(1)]),
    dependencies: (0, schema_utils_1.LooseRecord)(zod_1.z.object({ version: zod_1.z.string() })),
})
    .transform(({ lockfileVersion, dependencies: packages }) => ({
    lockfileVersion,
    packages,
}));
exports.PackageLock = schema_utils_1.Json.pipe(zod_1.z.union([exports.PackageLockV3Schema, exports.PackageLockPreV3Schema])).transform(({ packages, lockfileVersion }) => {
    const lockedVersions = {};
    for (const [entry, val] of Object.entries(packages)) {
        lockedVersions[entry] = val.version;
    }
    return { lockedVersions, lockfileVersion };
});
