// renovate/dist/util/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Result = void 0;
class Result {
    res;
    static ok(value) {
        return new Result({ success: true, value });
    }
    static err(e) {
        if (typeof e === 'undefined' && arguments.length === 0) {
            return new Result({ success: false, error: true });
        }
        const error = e;
        return new Result({ success: false, error });
    }
    static wrapCallback(callback) {
        try {
            return Result.ok(callback());
        }
        catch (error) {
            return Result.err(error);
        }
    }
    static wrapPromise(promise) {
        return promise.then((value) => Result.ok(value), (error) => Result.err(error));
    }
    static wrap(input) {
        return input instanceof Promise
            ? Result.wrapPromise(input)
            : Result.wrapCallback(input);
    }
    constructor(res) {
        this.res = res;
    }
    transform(fn) {
        return this.res.success
            ? Result.ok(fn(this.res.value))
            : Result.err(this.res.error);
    }
    catch(fallback) {
        return this.res.success ? this.res.value : fallback;
    }
    get value() {
        return this.res.success ? this.res.value : undefined;
    }
    get error() {
        return this.res.success ? undefined : this.res.error;
    }
}
exports.Result = Result;
