// renovate/dist/config/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = exports.getDefault = void 0;
const options_1 = require("./options");
// Use functions instead of direct values to avoid introducing global references.
// In particular, we want a new array instance every time we request a default array
// instead of sharing a single instance - mutation of this value could cause serious problems.
// See https://github.com/mend/renovate-on-prem/issues/290 for an example
const defaultValueFactories = {
    boolean: () => true,
    array: () => [],
    string: () => null,
    object: () => null,
    integer: () => null,
};
function getDefault(option) {
    return option.default === undefined
        ? defaultValueFactories[option.type]()
        : option.default;
}
exports.getDefault = getDefault;
function getConfig() {
    const options = (0, options_1.getOptions)();
    const config = {};
    options.forEach((option) => {
        if (!option.parent) {
            config[option.name] = getDefault(option);
        }
    });
    return config;
}
exports.getConfig = getConfig;
